File: app/models/investment.py
"""
Modèle de base de données pour les investissements.
"""
from decimal import Decimal
from enum import Enum as PyEnum
from typing import List, Optional

from sqlalchemy import (
    Column, String, Date, Numeric, Boolean, Text, 
    ForeignKey, Enum as SQLEnum, CheckConstraint
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from app.domain.entities.investment import Investment
from app.domain.value_objects.money import Money
from app.domain.value_objects.percentage import Percentage

from .base import BaseModel, GUID


class AssetType(PyEnum):
    """Types d'actifs supportés."""
    STOCK = "stock"
    ETF = "etf"
    CRYPTO = "crypto"
    BOND = "bond"
    COMMODITY = "commodity"
    REIT = "reit"
    MUTUAL_FUND = "mutual_fund"


class MarketCapCategory(PyEnum):
    """Catégories de capitalisation boursière."""
    LARGE_CAP = "large_cap"      # > $10B
    MID_CAP = "mid_cap"          # $2B - $10B
    SMALL_CAP = "small_cap"      # $300M - $2B
    MICRO_CAP = "micro_cap"      # < $300M


class DBInvestment(BaseModel):
    """Modèle investissement."""
    
    __tablename__ = "investments"
    
    # Relations
    user_id = Column(
        GUID,
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
        comment="ID de l'utilisateur propriétaire"
    )
    
    # Informations de base
    symbol = Column(
        String(20),
        nullable=False,
        comment="Symbole de l'investissement (ex: AAPL)"
    )
    
    name = Column(
        String(255),
        nullable=False,
        comment="Nom de l'entreprise/actif"
    )
    
    # Classification
    asset_type = Column(
        SQLEnum(AssetType),
        nullable=False,
        comment="Type d'actif"
    )
    
    country = Column(
        String(3),
        nullable=False,
        comment="Code pays ISO 3166-1 alpha-3"
    )
    
    sector = Column(
        String(100),
        nullable=True,
        comment="Secteur d'activité"
    )
    
    industry = Column(
        String(100),
        nullable=True,
        comment="Industrie spécifique"
    )
    
    market_cap_category = Column(
        SQLEnum(MarketCapCategory),
        nullable=True,
        comment="Catégorie de capitalisation boursière"
    )
    
    # Informations d'achat
    purchase_date = Column(
        Date,
        nullable=False,
        comment="Date d'achat"
    )
    
    purchase_price = Column(
        Numeric(15, 4),
        nullable=False,
        comment="Prix d'achat unitaire"
    )
    
    quantity = Column(
        Numeric(15, 8),
        nullable=False,
        comment="Quantité détenue"
    )
    
    currency = Column(
        String(3),
        nullable=False,
        default="USD",
        comment="Devise de l'investissement (ISO 4217)"
    )
    
    @property
    def total_cost(self):
        """Calcule le coût total d'achat."""
        return self.purchase_price * self.quantity

    # Informations supplémentaires
    dividend_yield = Column(
        Numeric(5, 4),
        nullable=True,
        comment="Rendement de dividende"
    )
    
    expense_ratio = Column(
        Numeric(5, 4),
        nullable=True,
        comment="Ratio de frais (pour ETF/fonds)"
    )
    
    notes = Column(
        Text,
        nullable=True,
        comment="Notes personnelles"
    )
    
    # Statut
    is_active = Column(
        Boolean,
        nullable=False,
        default=True,
        comment="Indique si l'investissement est actif"
    )
    
    # Relations
    user = relationship(
        "User",
        back_populates="investments",
        lazy="select"
    )
    
    price_history = relationship(
        "PriceHistory",
        back_populates="investment",
        cascade="all, delete-orphan",
        lazy="select",
        order_by="PriceHistory.timestamp.desc()"
    )
    
    transactions = relationship(
        "InvestmentTransaction",
        back_populates="investment",
        cascade="all, delete-orphan",
        lazy="select",
        order_by="InvestmentTransaction.transaction_date.desc()"
    )
    
    # Contraintes
    __table_args__ = (
        CheckConstraint('quantity > 0', name='chk_positive_quantity'),
        CheckConstraint('purchase_price > 0', name='chk_positive_purchase_price'),
        CheckConstraint('purchase_date <= CURRENT_DATE', name='chk_purchase_date_not_future'),
    )
    
    def __repr__(self) -> str:
        return f"<Investment(id={self.id}, symbol='{self.symbol}', name='{self.name}', user_id={self.user_id})>"
    
    def to_domain(self) -> Investment:
        return Investment(
            id=self.id,
            symbol=self.symbol,
            name=self.name,
            user_id=self.user_id,
            asset_type=self.asset_type,
            country=self.country,
            purchase_date=self.purchase_date,
            purchase_price=Money(amount=self.purchase_price, currency=self.currency),
            quantity=self.quantity,
            currency=self.currency,
            sector=self.sector,
            industry=self.industry,
            market_cap_category=self.market_cap_category,
            dividend_yield=Percentage(self.dividend_yield) if self.dividend_yield is not None else None,
            expense_ratio=Percentage(self.expense_ratio) if self.expense_ratio is not None else None,
            notes=self.notes,
            is_active=self.is_active
        )

File: mobile app/lib/models/investment.dart
double _toDouble(dynamic value, [double fallback = 0]) {
  if (value == null) return fallback;
  if (value is num) return value.toDouble();
  if (value is String) return double.tryParse(value) ?? fallback;
  return fallback;
}

class Investment {
  final String id;
  final String userId;
  final String symbol;
  final String name;
  final String assetType;
  final String country;
  final String? sector;
  final String? industry;
  final String? marketCapCategory;
  final String purchaseDate;
  final double purchasePrice;
  final double quantity;
  final String currency;
  final double? dividendYield;
  final double? expenseRatio;
  final String? notes;
  final String? investmentThesis;
  final String? thesisStatus;
  final double? alertThresholdPct;
  final String? accountType;
  final bool isActive;
  final String createdAt;
  final String updatedAt;

  Investment({
    required this.id,
    required this.userId,
    required this.symbol,
    required this.name,
    required this.assetType,
    required this.country,
    this.sector,
    this.industry,
    this.marketCapCategory,
    required this.purchaseDate,
    required this.purchasePrice,
    required this.quantity,
    required this.currency,
    this.dividendYield,
    this.expenseRatio,
    this.notes,
    this.investmentThesis,
    this.thesisStatus,
    this.alertThresholdPct,
    this.accountType,
    required this.isActive,
    required this.createdAt,
    required this.updatedAt,
  });

  double get totalCost => purchasePrice * quantity;

  factory Investment.fromJson(Map<String, dynamic> json) {
    return Investment(
      id: json['id'] as String,
      userId: json['user_id'] as String,
      symbol: json['symbol'] as String,
      name: json['name'] as String,
      assetType: json['asset_type'] as String,
      country: json['country'] as String? ?? '',
      sector: json['sector'] as String?,
      industry: json['industry'] as String?,
      marketCapCategory: json['market_cap_category'] as String?,
      purchaseDate: json['purchase_date'] as String,
      purchasePrice: _toDouble(json['purchase_price']),
      quantity: _toDouble(json['quantity']),
      currency: json['currency'] as String? ?? 'USD',
      dividendYield: json['dividend_yield'] != null
          ? _toDouble(json['dividend_yield'])
          : null,
      expenseRatio: json['expense_ratio'] != null
          ? _toDouble(json['expense_ratio'])
          : null,
      notes: json['notes'] as String?,
      investmentThesis: json['investment_thesis'] as String?,
      thesisStatus: json['thesis_status'] as String?,
      alertThresholdPct: json['alert_threshold_pct'] != null
          ? _toDouble(json['alert_threshold_pct'])
          : null,
      accountType: json['account_type'] as String?,
      isActive: json['is_active'] as bool? ?? true,
      createdAt: json['created_at'] as String,
      updatedAt: json['updated_at'] as String,
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'user_id': userId,
    'symbol': symbol,
    'name': name,
    'asset_type': assetType,
    'country': country,
    'sector': sector,
    'industry': industry,
    'market_cap_category': marketCapCategory,
    'purchase_date': purchaseDate,
    'purchase_price': purchasePrice,
    'quantity': quantity,
    'currency': currency,
    'dividend_yield': dividendYield,
    'expense_ratio': expenseRatio,
    'notes': notes,
    'investment_thesis': investmentThesis,
    'thesis_status': thesisStatus,
    'alert_threshold_pct': alertThresholdPct,
    'account_type': accountType,
    'is_active': isActive,
    'created_at': createdAt,
    'updated_at': updatedAt,
  };
}

class InvestmentCreate {
  final String accountType;
  final String? tickerSymbol;
  final String? isin;
  final String purchaseDate;
  final double quantity;
  final String? notes;
  final String? investmentThesis;
  final String? thesisStatus;
  final double? alertThresholdPct;

  InvestmentCreate({
    required this.accountType,
    this.tickerSymbol,
    this.isin,
    required this.purchaseDate,
    required this.quantity,
    this.notes,
    this.investmentThesis,
    this.thesisStatus,
    this.alertThresholdPct,
  });

  Map<String, dynamic> toJson() {
    final map = <String, dynamic>{
      'account_type': accountType,
      'purchase_date': purchaseDate,
      'quantity': quantity,
    };
    if (tickerSymbol != null) map['ticker_symbol'] = tickerSymbol;
    if (isin != null) map['isin'] = isin;
    if (notes != null && notes!.isNotEmpty) map['notes'] = notes;
    if (investmentThesis != null && investmentThesis!.isNotEmpty) {
      map['investment_thesis'] = investmentThesis;
    }
    if (thesisStatus != null) map['thesis_status'] = thesisStatus;
    if (alertThresholdPct != null) map['alert_threshold_pct'] = alertThresholdPct;
    return map;
  }
}

class InvestmentInitialValues {
  final String? accountType;
  final String? tickerSymbol;
  final double? suggestedQuantity;
  final String? investmentThesis;
  final String? notes;
  final double? alertThresholdPct;

  const InvestmentInitialValues({
    this.accountType,
    this.tickerSymbol,
    this.suggestedQuantity,
    this.investmentThesis,
    this.notes,
    this.alertThresholdPct,
  });
}
